services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -0,0 +1,227 @@
+    } else {
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_UNLIKELY_ERROR;
+    }
+}
+
+void CurrentTimeService::onCurrentTimeWritten(GattWriteAuthCallbackParams *write_request)
+{
+    CurrentTime input_time(write_request->data);
+
+    if (!input_time.valid()) {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_OUT_OF_RANGE;
+        return;
+    }
+
+    struct tm remote_time_tm{};
+    input_time.to_tm(&remote_time_tm);
+    time_t remote_time = mktime(&remote_time_tm);
+
+    set_time(remote_time, input_time.adjust_reason);
+
+    if (_current_time_handler) {
+        _current_time_handler->on_current_time_changed(remote_time, input_time.adjust_reason);
+    }
+
+    if (input_time.fractions256) {
+        write_request->authorizationReply = (GattAuthCallbackReply_t) DATA_FIELD_IGNORED;
+    } else {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    }
+}
+
+CurrentTimeService::CurrentTime::CurrentTime(const uint8_t *data)
+{
+    // FIXME: 2bit transmitted in little endian
+    year          = *data | (*(data + 1) << 8);
+    month         = *data++;
+    day           = *data++;
+    hours         = *data++;
+    minutes       = *data++;
+    seconds       = *data++;
+    weekday       = *data++;
+    fractions256  = *data++;
+    adjust_reason = *data;
+}
+
+CurrentTimeService::CurrentTime::CurrentTime(const struct tm *local_time_tm)
+{
+    // FIXME: tm_year should be coded in little endian
+    year          = (local_time_tm->tm_year + 1900);
+    month         =  local_time_tm->tm_mon  + 1;
+    day           =  local_time_tm->tm_mday;
+    hours         =  local_time_tm->tm_hour;
+    minutes       =  local_time_tm->tm_min;
+    seconds       =  local_time_tm->tm_sec;
+    /*
+     * The tm_wday field of a tm struct means days since Sunday (0-6)
+     * However, the weekday field of a CurrentTime struct means Mon-Sun (1-7)
+     * So, if tm_wday = 0, i.e. Sunday, the correct value for weekday is 7
+     * Otherwise, the fields signify the same days and no correction is needed
+     * */
+    weekday       =  local_time_tm->tm_wday == 0 ? 7 : local_time_tm ->tm_wday;
+    fractions256  =  0;
+    adjust_reason =  0;
+}
+
+
+bool CurrentTimeService::CurrentTime::valid()
+{
+    if ((year    < 1582) || (year    > 9999)) {
+        return false;
+    }
+    if ((month   <    1) || (month   >   12)) {
+        return false;
+    }
+    if ((day     <    1) || (day     >   31)) {
+        return false;
+    }
+    if ( hours   >   23) {
+        return false;
+    }
+    if ( minutes >   59) {
+        return false;
+    }
+    if ( seconds >   59) {
+        return false;
+    }
+    if ((weekday <    1) || (weekday >    7)) {
+        return false;
+    }
+
+    return true;
+}
+
+bool CurrentTimeService::CurrentTime::to_tm(struct tm * remote_time_tm)
+{
+    if (!valid()) {
+        return false;
+    }
+
+    remote_time_tm->tm_year  =  year  - 1900;
```

Comment:
you cannot read the year directly, it needs to be deserialised

Reply:
It's deserialised in the constructor

Reply:
it should not be

Reply:
oh, wait, you're not sending this out

Reply:
I'm not sending it out. The data is passed in by the client and a pointer to the data steam is used to construct a `CurrentTime` object in `onCurrentTimeRead()` The constructor deserialises the data. This function simply converts a `CurrentTime` object to libc `tm` struct.

Reply:
line 108

Reply:
This thread began with a comment about the `to_tm` member which is called in on data written.

Reply:
Let me explain this:
you have a struct that has multibyte values
you also use the same struct a buffer which you then send verbatim over the air
the value needs to be stored therefore in network order
at the same time you read that value in your code on your host, this means every time you read this value stored in network order you need to convert it to host order
in most cases on arm host is little endian same as network order so this is a nop
on a big endian this will break